sway-lsp/src/traverse/typed_tree.rs

```diff
@@ -233,7 +233,7 @@ impl<'a> TypedTree<'a> {
                 // impl is registered
                 self.collect_type_id(
                     implementing_for.type_id,
-                    &TypedAstToken::TypedDeclaration(declaration.clone()),
+                    &TypedAstToken::TypedArgument(implementing_for.clone()),
```

Comment:
This change apparently breaks some of the struct code actions. Specifically, `all_references_of_token` used to return the `TypedAstToken::TypedDeclaration(TyDecl::ImplTrait)` given a struct decl. For example, the reason this test is failing:

```
struct A {
    a: u64,
    b: u64,
}

impl A {
    fn f(ref mut self) {
        self.a = 42;
        self.b = 77;
    }
}
```

We check if there are any existing impls for `A`, to see if we can insert the generated `new` function inside the existing impl ([code](https://github.com/FuelLabs/sway/blob/4e0291c48b23e047bbfab5a52ee4c4842e6cc1a6/sway-lsp/src/capabilities/code_actions/struct_decl/struct_new.rs#L21-L33)).

What's happening now seems to be that the token for the `TyImpl` is not getting inserted into the token map, so `all_references_of_token` is not returning it and we can't find it. The result is that the code action can only generate a `new` function inside of a new impl block, which causes the tests to fail.

I'm not sure of the best way to fix this - I see that renaming impls works on your branch, so maybe we could use the same mechanism for finding existing impls in `struct_new.rs` to make it work for code actions as well?

Reply:
Here's the difference:

failing tests

<img width="515" alt="image" src="https://user-images.githubusercontent.com/47993817/231055639-48d2dada-31c4-4daf-97f7-fc585733f099.png">

working (by changing this line back)

<img width="531" alt="image" src="https://user-images.githubusercontent.com/47993817/231055865-6a7648e6-9bb1-4368-bea5-9c8158e6deaf.png">

Reply:
Does changing this back break renaming?

Reply:
thanks for looking into this! I've got the tests working again now. 

We now pass in a different typed_token in the `collect_type_id` depending on if `trait_decl_ref.is_some()`. If it isn't (in the case of a struct impl) then we pass through the original decl which contains the `ImplTrait`. Otherwise, we mark this token a `TypeArgument`.

So if we take the following:

```rust
abi MyContract {
    fn test_function();
}

impl MyContract for Contract {
    fn test_function() {}
}
```

In this case the `MyContact` token is the declaration, while the `Contract` token is the `TypeArgument` token. This allows us to one have one Declaration for the code block which is needed for the renaming logic, plus I think it makes more sense conceptually as well.